Added support for getting the mode and also battery_level from device

viswa-swami opened the pull request:
Currrent code didn't have the capability to read the mode from the arlo base station/device.  Added support to do the same.  Also added property code to get the battery level of all the cameras (as we are talking to base station), and similarly added properties to print out the camera properties, base station properties, schedule, rules etc.

Tested the code in my home arlo system and it works.  Once this is approved and signed-off, we can update the Home Assistant code (which uses this pyarlo library) to fetch the status from the device directly.

This will help in keeping sync between app, website and Hass.

We basically needed to subscribe ourself (python app using this library is the client) with the base station, and then subscribe for notifications. Then when client requests for modes or schedules or rules etc, we would get the notification via the event stream.  We need to parse that and return the data.

Fixes: https://github.com/tchellomello/python-arlo/issues/20 https://github.com/tchellomello/python-arlo/issues/5

tchellomello commented:
Awesome @viswa-swami!! I'll review and test it tonight!!  Great work!!!

viswa-swami commented:
Test code:
```
from pyarlo import PyArlo
Arlo=PyArlo(username, password)
Base=arlo.base_stations[0]
print(Base.mode)
print(Base.get_camera_battery_level)
```

viswa-swami commented:
I will update the README with these newly added properties later today or tomorrow.

tchellomello commented on pyarlo/__init__.py:
```diff
@@ -90,7 +90,8 @@ def query(self,
               retry=3,
-              raw=False):
+              raw=False,
+              stream=False):
```
@viswa-swami  can you please add stream on the doc-string documentation?

viswa-swami replied:
Done.

tchellomello commented on pyarlo/base_station.py:
```diff
@@ -1,15 +1,20 @@
+import sseclient
+from pyarlo.const import ACTION_MODES, NOTIFY_ENDPOINT, RESOURCES
+from pyarlo.const import (
+    ACTION_BODY, SUBSCRIBE_ENDPOINT, UNSUBSCRIBE_ENDPOINT)
```
let's make just one line to include all constants
```python
from pyarlo.const import (
  ACTION_BODY, ACTION_MODES, NOTIFY_ENDPOINT, 
  RESOURCES, SUBSCRIBE_ENDPOINT, UNSUBSCRIBE_ENDPOINT)
```

viswa-swami replied:
Done.

tchellomello commented on pyarlo/base_station.py:
```diff
@@ -1,15 +1,20 @@
     """Arlo Base Station module implementation."""
 
-    def __init__(self, name, attrs, arlo_session):
+    def __init__(self, name, attrs, session_token, arlo_session):
```
can add the session_token on the doc string?

viswa-swami replied:
Done.

tchellomello commented on pyarlo/base_station.py:
```diff
@@ -19,11 +24,144 @@ def __init__(self, name, attrs, arlo_session):
+                break
+            data = json.loads(event.data)
+            if 'status' in data:
+                if data['status'] == "connected":
```
try to use data.get('status') to avoid tracebacks if/when status is not part of data

viswa-swami replied:
Done.

tchellomello commented on pyarlo/base_station.py:
```diff
@@ -19,11 +24,144 @@ def __init__(self, name, attrs, arlo_session):
+            if 'status' in data:
+                if data['status'] == "connected":
+                    _LOGGER.debug("Successfully subscribed this base station")
+            elif 'action' in data:
```
Would be possible the  `data` dictionary to have `status` and `action` at the same time? should the `status` continue as connected and then have some `action`?

viswa-swami replied:
Status and output are not available at the same time, in the event stream output.

'status' is available only on the connection response that we get first.  After that, all the notifications have only the 'action' in the event stream.

That's why I added the 'status' and 'action' as if-else statement

tchellomello commented on pyarlo/base_station.py:
```diff
@@ -19,11 +24,144 @@ def __init__(self, name, attrs, arlo_session):
+            resource=None,
+            mode=None,
+            publish_response=None):
+        """Run action."""
```
@viswa-swami  I know that the older method did not have the doc string, but could you include the doc string now?

viswa-swami replied:
Done.

tchellomello commented on pyarlo/base_station.py:
```diff
@@ -113,7 +301,11 @@ def mode(self, mode):
+            method='SET',
+            resource='modes',
+            mode=mode,
+            publish_response=True)
```
make sure to call `update()` method after changing the mode otherwise it will not update. For example:

```python
In [4]: base.mode
Out[4]: u'armed'

In [5]: base.available_modes
Out[5]: ['schedule', 'armed', 'disarmed', 'custom']

In [6]: base.mode = 'disarmed'

In [7]: base.mode

In [8]: base.update()

In [9]: base.mode
Out[9]: u'disarmed'
```

viswa-swami replied:
If really required, we can call them in the mode setter function after running the __run_action() to be sure.

That way, we don't have to do that update explicitly.  When someone sets the mode, it gets updated automatically by the library itself.

viswa-swami replied:
I have anyway called the update() function all doing the run_action() just to be sure.  I just wanted to avoid returning the wrong status !

chaddotson commented:
I will also test with 3.6 once available.

viswa-swami commented:
I actually used python 3.5 to implement these changes and didn't see this error... !!! Let me try with 3.6 later today.

Not sure what that error is, just from the snapshot.

viswa-swami commented:
I actually tried to callback the base station function from the ArloCamera class and tried the exact same thing that you have requested for each camera, but somehow I was always getting error when I called the base station function.

Can you give an example or some reference as to how to call the base station function from camera class?

viswa-swami commented:
@tchellomello I tried with python3.6 and I am not getting the error like you did !!! Not sure if your requests module has some problem ?

```
$ python3.6
Python 3.6.1+ (default, Jun  8 2017, 12:24:27) 
[GCC 5.4.1 20170519] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from pyarlo import PyArlo
>>> arlo = PyArlo('viswa.swami@gmail.com', 'A!s2h3o4k5')
>>> base = arlo.base_stations[0]
>>> print (base.mode)
disarmed
>>> print (base.get_camera_battery_level)
{'4N71687T106E4': 92, '4N71687Y0D81E': 90}
```

viswa-swami commented:
@tchellomello Only thing remaining is adding the 'get_battery_level' for camera module.  I am not sure, but I am getting an error when trying to invoke the call from BaseStation in the Camera module.  I am not able to figure it out !!!!!  Can you give it a try !

The base.get_camera_battery_level() will return a dictionary, which we can filter based on camera's device_id and return only that battery level.

viswa-swami commented:
@tchellomello Figured out the mistake as to why the camera module was not able to get the battery levels.  I was making a mistake when accessing the dictionary.

I have added the code to get the battery level of each camera.  I will update the PR shortly in few minutes.

viswa-swami commented:
example code:

```
#!/usr/bin/python3.6

from pyarlo import PyArlo
arlo = PyArlo('viswa.swami@gmail.com', 'A!s2h3o4k5')
base = arlo.base_stations[0]
print (base.mode)
print (base.get_camera_battery_level)
cam = arlo.cameras[0]
print (cam.get_battery_level)
cam = arlo.cameras[1]
print (cam.get_battery_level)
print (base.get_camera_properties)
print (base.get_basestation_properties)
print (base.get_camera_rules)
```

viswa-swami commented:
@tchellomello Updated the code in the PR.  Added new properties for base station and camera. 

Added property in base station to fetch the schedule and a boolean property to find out if motion detection is enabled or not. 
Added properties in camera to get flip, mirror, power save mode, connection status, signal strength etc.

viswa-swami commented:
Test Code:

```
#!/usr/bin/python3.6

from pyarlo import PyArlo
arlo = PyArlo('viswa.swami@gmail.com', 'A!s2h3o4k5')
base = arlo.base_stations[0]
print ("Base: Active Mode is           : " + str(base.mode))
motion_enabled = base.is_motion_detection_enabled
print ("Base: Motion Detection Enabled : ", end="", flush=True)
if motion_enabled:
    print ("Yes")
else:
    print ("No")
print ("Base: Camera Battery Levels    : ", end="", flush=True)
print (base.get_camera_battery_level)
cam = arlo.cameras[0]
connection = cam.is_camera_connected
print ("Camera-1 Is it Connected?      : ", end="", flush=True)
if connection:
    print ("Yes")
else:
    print ("No")
print ("Camera-1 Battery Level is      : " + str(cam.get_battery_level) + "%")
print ("Camera-1 Signal Strength is    : " + str(cam.get_signal_strength))
print ("Camera-1 Brightness is         : " + str(cam.get_brightness))
print ("Camera-1 Flip State is         : " + str(cam.get_flip_state))
print ("Camera-1 Mirror State is       : " + str(cam.get_mirror_state))
print ("Camera-1 Power Save Mode is    : " + str(cam.get_powersave_mode))
cam = arlo.cameras[1]
connection = cam.is_camera_connected
print ("Camera-2 Is it Connected?      : ", end="", flush=True)
if connection:
    print ("Yes")
else:
    print ("No")
print ("Camera-2 Battery Level is      : " + str(cam.get_battery_level) + "%")
print ("Camera-2 Signal Strength is    : " + str(cam.get_signal_strength))
print ("Camera-2 Brightness is         : " + str(cam.get_brightness))
print ("Camera-2 Flip State is         : " + str(cam.get_flip_state))
print ("Camera-2 Mirror State is       : " + str(cam.get_mirror_state))
print ("Camera-2 Power Save Mode is    : " + str(cam.get_powersave_mode))
print ("Base: Camera Properties        : ")
print (base.get_camera_properties)
print ("Base: BaseStation Properties   : ")
print (base.get_basestation_properties)
print ("Base: Camera Rules             : ")
print (base.get_camera_rules)
print ("Base: Camera Schedule          : ")
print (base.get_camera_schedule)
```

viswa-swami commented:
Test Code:

#!/usr/bin/python3.6

from pyarlo import PyArlo
arlo = PyArlo('username', 'password')
base = arlo.base_stations[0]
print ("Base: Active Mode is           : " + str(base.mode))
motion_enabled = base.is_motion_detection_enabled
print ("Base: Motion Detection Enabled : ", end="", flush=True)
if motion_enabled:
    print ("Yes")
else:
    print ("No")
print ("Base: Camera Battery Levels    : ", end="", flush=True)
print (base.get_camera_battery_level)
cam = arlo.cameras[0]
connection = cam.is_camera_connected
print ("Camera-1 Is it Connected?      : ", end="", flush=True)
if connection:
    print ("Yes")
else:
    print ("No")
print ("Camera-1 Battery Level is      : " + str(cam.get_battery_level) + "%")
print ("Camera-1 Signal Strength is    : " + str(cam.get_signal_strength))
print ("Camera-1 Brightness is         : " + str(cam.get_brightness))
print ("Camera-1 Flip State is         : " + str(cam.get_flip_state))
print ("Camera-1 Mirror State is       : " + str(cam.get_mirror_state))
print ("Camera-1 Power Save Mode is    : " + str(cam.get_powersave_mode))
cam = arlo.cameras[1]
connection = cam.is_camera_connected
print ("Camera-2 Is it Connected?      : ", end="", flush=True)
if connection:
    print ("Yes")
else:
    print ("No")
print ("Camera-2 Battery Level is      : " + str(cam.get_battery_level) + "%")
print ("Camera-2 Signal Strength is    : " + str(cam.get_signal_strength))
print ("Camera-2 Brightness is         : " + str(cam.get_brightness))
print ("Camera-2 Flip State is         : " + str(cam.get_flip_state))
print ("Camera-2 Mirror State is       : " + str(cam.get_mirror_state))
print ("Camera-2 Power Save Mode is    : " + str(cam.get_powersave_mode))
print ("Base: Camera Properties        : ")
print (base.get_camera_properties)
print ("Base: BaseStation Properties   : ")
print (base.get_basestation_properties)
print ("Base: Camera Rules             : ")
print (base.get_camera_rules)
print ("Base: Camera Schedule          : ")
print (base.get_camera_schedule)

viswa-swami commented:
Test Code:

```
#!/usr/bin/python3.6

from pyarlo import PyArlo
arlo = PyArlo('username', 'password')
base = arlo.base_stations[0]
print ("Base: Active Mode is           : " + str(base.mode))
motion_enabled = base.is_motion_detection_enabled
print ("Base: Motion Detection Enabled : ", end="", flush=True)
if motion_enabled:
    print ("Yes")
else:
    print ("No")
print ("Base: Camera Battery Levels    : ", end="", flush=True)
print (base.get_camera_battery_level)
cam = arlo.cameras[0]
connection = cam.is_camera_connected
print ("Camera-1 Is it Connected?      : ", end="", flush=True)
if connection:
    print ("Yes")
else:
    print ("No")
print ("Camera-1 Battery Level is      : " + str(cam.get_battery_level) + "%")
print ("Camera-1 Signal Strength is    : " + str(cam.get_signal_strength))
print ("Camera-1 Brightness is         : " + str(cam.get_brightness))
print ("Camera-1 Flip State is         : " + str(cam.get_flip_state))
print ("Camera-1 Mirror State is       : " + str(cam.get_mirror_state))
print ("Camera-1 Power Save Mode is    : " + str(cam.get_powersave_mode))
cam = arlo.cameras[1]
connection = cam.is_camera_connected
print ("Camera-2 Is it Connected?      : ", end="", flush=True)
if connection:
    print ("Yes")
else:
    print ("No")
print ("Camera-2 Battery Level is      : " + str(cam.get_battery_level) + "%")
print ("Camera-2 Signal Strength is    : " + str(cam.get_signal_strength))
print ("Camera-2 Brightness is         : " + str(cam.get_brightness))
print ("Camera-2 Flip State is         : " + str(cam.get_flip_state))
print ("Camera-2 Mirror State is       : " + str(cam.get_mirror_state))
print ("Camera-2 Power Save Mode is    : " + str(cam.get_powersave_mode))
print ("Base: Camera Properties        : ")
print (base.get_camera_properties)
print ("Base: BaseStation Properties   : ")
print (base.get_basestation_properties)
print ("Base: Camera Rules             : ")
print (base.get_camera_rules)
print ("Base: Camera Schedule          : ")
print (base.get_camera_schedule)

```

viswa-swami commented:
After your review/test, If you are ok with these changes, then I can merge these PR changes into the master.....

tchellomello commented:
@viswa-swami  awesome work!! I'll test it tonight :1st_place_medal:

viswa-swami commented:
Added new property to print out the camera motion detection sensitivity.  We can then add these properties to be displayed for each camera in Home Assistant as properties/attributes.  We can also add potentially sensors for these which can then be used for automations and stuff.

Updated Example code with support for Motion Sensitivity:

```
#!/usr/bin/python3.6

from pyarlo import PyArlo
arlo = PyArlo('username', 'password')
base = arlo.base_stations[0]
print ("Base: Active Mode is           : " + str(base.mode))
motion_enabled = base.is_motion_detection_enabled
print ("Base: Motion Detection Enabled : ", end="", flush=True)
if motion_enabled:
    print ("Yes")
else:
    print ("No")
print ("Base: Camera Battery Levels    : ", end="", flush=True)
print (base.get_camera_battery_level)
cam = arlo.cameras[0]
connection = cam.is_camera_connected
print ("Camera-1 Is it Connected?      : ", end="", flush=True)
if connection:
    print ("Yes")
else:
    print ("No")
print ("Camera-1 Battery Level is      : " + str(cam.get_battery_level) + "%")
print ("Camera-1 Signal Strength is    : " + str(cam.get_signal_strength))
print ("Camera-1 Brightness is         : " + str(cam.get_brightness))
print ("Camera-1 Flip State is         : " + str(cam.get_flip_state))
print ("Camera-1 Mirror State is       : " + str(cam.get_mirror_state))
print ("Camera-1 Power Save Mode is    : " + str(cam.get_powersave_mode))
print ("Camera-1 Motion Sensitivity    : " + str(cam.get_motion_detection_sensitivity))
cam = arlo.cameras[1]
connection = cam.is_camera_connected
print ("Camera-2 Is it Connected?      : ", end="", flush=True)
if connection:
    print ("Yes")
else:
    print ("No")
print ("Camera-2 Battery Level is      : " + str(cam.get_battery_level) + "%")
print ("Camera-2 Signal Strength is    : " + str(cam.get_signal_strength))
print ("Camera-2 Brightness is         : " + str(cam.get_brightness))
print ("Camera-2 Flip State is         : " + str(cam.get_flip_state))
print ("Camera-2 Mirror State is       : " + str(cam.get_mirror_state))
print ("Camera-2 Power Save Mode is    : " + str(cam.get_powersave_mode))
print ("Camera-2 Motion Sensitivity    : " + str(cam.get_motion_detection_sensitivity))
print ("Base: Camera Properties        : ")
print (base.get_camera_properties)
print ("Base: BaseStation Properties   : ")
print (base.get_basestation_properties)
print ("Base: Camera Rules             : ")
print (base.get_camera_rules)
print ("Base: Camera Schedule          : ")
print (base.get_camera_schedule)

```

viswa-swami commented:
@tchellomello Let me know once you have reviewed and tested it.  We can merge it and publish to PIP.

That way, we can upgrade the dependency in HASS and add these attributes/properties to Arlo component in HASS and also sensors if appropriate.

viswa-swami commented:
@tchellomello Could you also please bump up the version to 0.0.5 and post it to PIP ? That way we can add the new version as requirements for HASS and add the properties there.

I have already added arm/disarm as a service to camera component in HASS (PR #7691), but need to add the support to get the mode actually from device.  That can be done once PyArlo gets published to PIP.

tchellomello commented:
@viswa-swami sure!! PIP package has been published!!

viswa-swami commented:
Thank you @tchellomello.  I will use this version for HASS and add the new properties/features in it also.
